File: Java/src/matrix.java
import java.io.BufferedReader;
import java.io.FileOutputStream;
import java.io.FileReader;
import java.io.IOException;

public class matrix{    
    // 定义矩阵类的变量
    int row;
    int col;
    int[][] value;

    public static void main(String[] args) throws IOException {
        System.out.printf("Matrix multiplication implemented by Java.\n");
        FileReader fr = new FileReader("data.txt");
        BufferedReader bf = new BufferedReader(fr);
        matrix A = new matrix(bf);
        matrix B = new matrix(bf);
        if (A.col != B.row) {
            System.out.print("Input Error");
            return;
        }
        // PrintArr(A.value);
        // PrintArr(B.value);
        long t = System.currentTimeMillis();
        matrix ans = A.Comput(B);
        t = System.currentTimeMillis() - t;
        System.out.printf("Matrix %d×%d and %d×%d cost %d ms.\n",A.row,A.col,B.row,B.col,t);
        // PrintArr(ans.value);
        write2file(ans);
        fr.close();
        bf.close();
    }
    // 读取矩阵
    public matrix(BufferedReader bf) throws IOException {
        // 按行读取
        String str;
        // 读取行列
        str = bf.readLine();
        String[] temp = str.split(" ");
        row = Integer.parseInt(temp[0]);
        col = Integer.parseInt(temp[1]);
        value = new int[row][col];
        for (int i = 0; i < row; i++) {
            str = bf.readLine();
            temp = str.split(" ");
            for (int j = 0; j < col; j++)
                value[i][j] = Integer.parseInt(temp[j]);
        }
        // PrintArr(value);
    }

    public static void write2file(matrix temp) throws IOException
    {
        FileOutputStream outputStream = new FileOutputStream("result.txt");
        String str;
        str = Integer.toString(temp.row)+" " + Integer.toString(temp.col) + "\n";
        for(int i = 0 ; i < temp.row ; i++)
        {
            for(int j = 0 ; j < temp.col ; j++)
            {
                str += Integer.toString(temp.value[i][j])+" ";
            }
            str += "\n";
        }
        byte[] strToBytes = str.getBytes();
        outputStream.write(strToBytes);
        outputStream.close();
    }

    public static void PrintArr(int[][] temp)
    {
        for(int i = 0 ; i < temp.length; i++)
        {
           for(int j = 0 ; j < temp[i].length ;j++)
           {
               System.out.print(temp[i][j]+" ");
           }
           System.out.print('\n');
        }
    }
    
    public matrix(int row,int col)
    {
        this.row = row;
        this.col = col;
        this.value = new int[row][col];
    }
    // 计算结果
    public matrix Comput(matrix B)
    {
        matrix Ans = new matrix(this.row,B.col);
        for(int i = 0 ; i < this.row ; i++)
        {
            for(int j = 0 ; j < B.col ; j++)
            {
                for(int times = 0 ; times < this.col ; times++)
                    Ans.value[i][j] += this.value[i][times]*B.value[times][j];
            }
        }
        return Ans;
    }
}

File: Python/src/matrix.py
#coding = utf-8
#owner: IFpop

from time import *

'''
# 设置文件数据格式为  n1 n2  行数 列数
#                   .....  n1行数据
# 共两个矩阵
# 返回值为： 矩阵，矩阵行列list
'''
def read_matrix(f):
    temp = []
    num =  [int(x) for x in (f.readline()).split(' ')]
    # print("矩阵n×n行数:")
    # print(num)
    # 按行读取
    for i in range(num[0]):
        temp_list = [int(x) for x in (f.readline()).split(' ')[:-1]]
        # print(temp_list)
        temp.append(temp_list)
    return temp,num

def Comput(A,A_num,B,B_num):
    Ans = [[0 for j in range(B_num[1])] for i in range(A_num[0])]
    for i in range(A_num[0]):
        for j in range(B_num[1]):
            for time in range(A_num[1]): # 前一个矩阵的列与后一个矩阵的行是相等的
                Ans[i][j] += A[i][time]*B[time][j]
    return Ans

get_now_milli_time = lambda: int(time() * 1000)

if __name__ == '__main__':
    print("Matrix multiplication implemented by Python.")
    with open('data.txt','r') as f:
        # 读取A矩阵
        A,A_num = read_matrix(f)
        # 读取B矩阵
        B,B_num = read_matrix(f)
        f.close()
    if(A_num[1] == B_num[0]):
        begin_time = get_now_milli_time()
        ans = Comput(A,A_num,B,B_num)
        end_time = get_now_milli_time()
        run_time = end_time-begin_time
        # 将ans写入文件
        with open('result.txt','w') as f:
            f.write(str(A_num[0])+" "+str(B_num[1])+"\n")
            for i in range(A_num[0]):
                for j in range(B_num[1]):
                    f.write(str(ans[i][j])+" ")
                f.write('\n')
        print(f'Matrix {A_num[0]} × {A_num[1]} and {B_num[0]} × {B_num[1]} cost time: {run_time} ms.\n')
    else:
        print("Row and column mismatch")
    